Add computeJacobian() and computeOffDiagJacobian() for NodalBCs

jwpeterson opened the pull request:
This branch adds the ability for `NodalBCs` to specify on- and off-diagonal Jacobian entries other than 1.0 and 0.0.  It also adds a test of the new capability which has a nonlinear `NodalBC` with off-diagonal coupling.  Given a bad enough initial guess, I've noticed that this test diverges with `solve_type=NEWTON` under the old code path:

```
NEWTON w/ SMP, full = true
--------------------------
 0 Nonlinear |R| = 3.075711e+01
 1 Nonlinear |R| = 2.819131e+02
 2 Nonlinear |R| = 1.888818e+04
 3 Nonlinear |R| = 1.104392e+08
 4 Nonlinear |R| = 3.678752e+15
*** solve fails ***
```

while it converges nicely under the new code path, which uses the correct Jacobian:

```
NEWTON w/ SMP, full = true
--------------------------
 0 Nonlinear |R| = 3.075711e+01
 1 Nonlinear |R| = 2.819131e+02
 2 Nonlinear |R| = 6.637843e+01
 3 Nonlinear |R| = 1.314614e+01
 4 Nonlinear |R| = 1.413857e+00
 5 Nonlinear |R| = 2.654542e-02
 6 Nonlinear |R| = 1.010187e-05
 7 Nonlinear |R| = 1.464852e-12
```

We have a milestone for this due 1/30/2015.

Refs #656.

jwpeterson committed:
Add computeJacobian(), computeOffDiagJacobian() for NodalBCs.

Refs #656.

jwpeterson committed:
Rename 'coupled_bc' test to 'matched_value_bc'

This better reflects what that test is actually for.

Refs #656.

jwpeterson committed:
Adding CoupledDirichletBC test.

This tests the new on- and off-diagonal Jacobians for NodalBCs.

Also fixing DirichletBCfuncXYZ0, which had an incorrect
computeQpJacobian() and removing DirichletBCfuncXYZ1, which was not
used by anything.

Refs #656.

jwpeterson committed:
Clean up one of the module input files, use timestep_end.

Refs #656.

jwpeterson committed:
Implement computeQpOffDiagJacobian() for MatchedValueBC.

This test can now use Newton and converge in a single nonlinear step,
rather than relying on PJFNK to get the correct action of the
Jacobian.

Refs #656.

moosebuild commented:
Results of testing aac206e7a39645b9304dd6060c4dd9567353d1cc using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10705

moosebuild commented:
Results of testing aac206e7a39645b9304dd6060c4dd9567353d1cc using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10706

moosebuild commented:
Results of testing aac206e7a39645b9304dd6060c4dd9567353d1cc using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10707

moosebuild commented:
Results of testing 091d5264541de078736c588fc72c5b99871670e5 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10717

moosebuild commented:
Results of testing 091d5264541de078736c588fc72c5b99871670e5 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10718

moosebuild commented:
Results of testing 091d5264541de078736c588fc72c5b99871670e5 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10719

friedmud commented:
Coolness!

friedmud commented on framework/include/base/Assembly.h:
```diff
@@ -672,6 +672,43 @@ class Assembly
+  /**
+   * Return a writeable reference to the cached Nodal BC data.
+   */
+  CachedNodalBCData & cachedNodalBCData() { return _cached_nodal_bc_data; }
```
I don't really understand the point of this - if it's just going to return a writeable reference - why not just make the member variable public?

I mean: I'm not 100% sure that's the right idea anyway - but this seems to just straddle the line...

andrsd replied:
IMO, there are 2 ways here:
  1. make the member variables in the class protected and add `CachedNodalBCData::addValue(i, j, value)`
  2. add `Assembly::cacheNodalBCValue(i, j, value)` and let the method do the caching.
   Change `CachedNodalBCData::clear()` to `Assembly::clearCachedNodalBCValues()`.
   How it is handled inside `Assembly` is no business of the caller and the caller does not have to get the writable reference. Unless, of course, there is a future plan that would require an access to `CachedNodalBCData`.

I'd go with (2).

friedmud commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -1772,11 +1772,20 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
+    // Compute Jacobians for NodalBCs
     ConstBndNodeRange & bnd_nodes = *_mesh.getBoundaryNodeRange();
-    for (ConstBndNodeRange::const_iterator nd = bnd_nodes.begin() ; nd != bnd_nodes.end(); ++nd)
+    for (ConstBndNodeRange::const_iterator nd = bnd_nodes.begin(); nd != bnd_nodes.end(); ++nd)
```
Seems like this should be able to be done in a ThreadedNodeLoop

You would have chached data from each thread that you would join down.

Not necessary on this first pass - but something to think about...

friedmud replied:
In fact - thinking more about this it should be trivial to implement because each thread has an `Assembly` where it would cache the values and then during join you would join them down from one thread to the next...

jwpeterson replied:
> On Jan 22, 2015, at 8:44 AM, Derek Gaston <notifications@github.com> wrote:
> 
> In framework/src/base/NonlinearSystem.C:
> 
> >      ConstBndNodeRange & bnd_nodes = *_mesh.getBoundaryNodeRange();
> > -    for (ConstBndNodeRange::const_iterator nd = bnd_nodes.begin() ; nd != bnd_nodes.end(); ++nd)
> > +    for (ConstBndNodeRange::const_iterator nd = bnd_nodes.begin(); nd != bnd_nodes.end(); ++nd)
> In fact - thinking more about this it should be trivial to implement because each thread has an Assembly where it would cache the values and then during join you would join them down from one thread to the next...
> 

It looked to me that all the BCs were on thread 0's BCwarehouse (from the original code).  But if they are on all threads then this should work fine...

friedmud replied:
Hmmm - that would be something to check.  I do think they're created for all threads - but we were only using thread 0's... but I could be wrong.  Shouldn't be tough to fix if it's not though...

friedmud commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -1786,16 +1795,65 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
       {
         _fe_problem.reinitNodeFace(node, boundary_id, 0);
```
I forget - do we know that boundary nodes have at least one boundary condition on them?  We should make sure that we're not calling this when there aren't any BCs on this node...

Note: I know this is the way it was before... just talking about it.

However: we _do_ have to do this before we call `shouldApply()`... so we should be careful...

andrsd replied:
> I forget - do we know that boundary nodes have at least one boundary condition on them?

I do not think that is guaranteed.

friedmud replied:
Then we should try to short-circuit this by making sure that there are _any_ BCs here before we do any work.

jwpeterson replied:
I think it is guaranteed.  `MooseMesh::getBoundaryNodeRange()` returns iterators into the `MooseMesh::_bnd_nodes` vector.  The `_bnd_nodes` vector is created by calling   `getMesh().get_boundary_info().build_node_list(nodes, ids)`, which only adds nodes which have BCs on them.

andrsd replied:
I thought that `boundary_info()` was created based on sidesets and nodesets. Then, just because there is a sideset, it does not mean there will be a MOOSE BC object on it. But, if the first sentence is not true, then disregard...

jwpeterson replied:
> just because there is a sideset, it does not mean there will be a MOOSE BC

Right, but doesn't that check, combined with this check:
```
if (_bcs[0].hasNodalBCs(boundary_id) && node->processor_id() == processor_id())
```
guarantee that that nodeset has NodalBCs?

andrsd replied:
Yes, this guards against it. I just did not see it in the patch context. So, all is good here as far as I can tell.

friedmud commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -1786,16 +1795,65 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
+            {
+              // Is the BC actually coupled to jvar?
+              bool bc_is_coupled_to_jvar = false;
+              for (unsigned int var = 0; var < coupled_moose_vars.size(); ++var)
```
Seems like we should process this first before we start the node loop and figure this out once - instead of doing it once per node.

friedmud commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -1786,16 +1795,65 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
+              // NodalBC is always coupled to its own variable) or the
+              // BC is coupled to jvar.
+              if (ivar==jvar || bc_is_coupled_to_jvar)
+              {
```
Extra curly braces :wink:

friedmud commented on framework/include/base/Assembly.h:
```diff
@@ -672,6 +672,43 @@ class Assembly
+    // Clears the cached data arrays
+    void clear()
+    {
+      cached_nodal_bc_vals.clear();
```
You may want to explicitly remember the `capacity()` of these vectors and call `reserve()` on them after clearing them.

`clear()` is not guaranteed to keep the capacity in tact... and you're using `push_back()` to fill these things in - so it may be a good idea to make sure that `push_back()` is only slow the first time these BCs are computed...

friedmud replied:
You can see where I did something similar in Assembly for cached residuals and Jacobians:

friedmud replied:
https://github.com/idaholab/moose/blob/devel/framework/src/base/Assembly.C#L1067

I keep track of the max size of the vectors I ever see - and make sure the vectors always have at least that much reserved (in fact I do 2x that just in case)

friedmud commented on framework/src/bcs/NodalBC.C:
```diff
@@ -47,7 +47,60 @@ NodalBC::computeResidual(NumericVector<Number> & residual)
+    Assembly::CachedNodalBCData & cached_data = _assembly.cachedNodalBCData();
+
+    // Cache (i,j,value)
+    cached_data.cached_nodal_bc_vals.push_back(cached_val);
```
Scary: Make sure these vectors have enough `capacity()` to keep re-allocs from happening (the first time it's probably unavoidable).

friedmud commented on framework/src/bcs/NodalBC.C:
```diff
@@ -47,7 +47,60 @@ NodalBC::computeResidual(NumericVector<Number> & residual)
+Real
+NodalBC::computeQpJacobian()
+{
+  return 1.;
```
I like this: good idea to keep backward compatibility....

friedmud commented on test/tests/bcs/coupled_dirichlet_bc/coupled_dirichlet_bc.i:
```diff
@@ -0,0 +1,110 @@
+  # Adding a CoupledForce term induces a coupling between the u and v
+  # equations.  Is this enough to trigger the off-diagonal Jacobian
+  # coupling?
+  [./coupled_force_u]
```
Is this required for the BCs to be able to do off-diagonal stuff?

andrsd replied:
It should not be. `CoupleableMooseVariableDependencyIntermediateInterface` (which NodalBCs do have) should take care of filling in the appropriate vector(s)...

friedmud replied:
Then - should we remove that from this test... or at least remove the comment?

The comment makes it sound like something is wrong...

jwpeterson replied:
> On Jan 22, 2015, at 8:46 AM, David Andrs <notifications@github.com> wrote:
> 
> In test/tests/bcs/coupled_dirichlet_bc/coupled_dirichlet_bc.i:
> 
> > +  [./v]
> > +    order = FIRST
> > +    family = LAGRANGE
> > +  [../]
> > +[]
> > +
> > +[Kernels]
> > +  [./diff_u]
> > +    type = Diffusion
> > +    variable = u
> > +  [../]
> > +
> > +  # Adding a CoupledForce term induces a coupling between the u and v
> > +  # equations.  Is this enough to trigger the off-diagonal Jacobian
> > +  # coupling?
> > +  [./coupled_force_u]
> It should not be. CoupleableMooseVariableDependencyIntermediateInterface (which NodalBCs do have) should take care of filling in the appropriate vector(s)...
> 

Yeah this comment is wrong... I will remove it, it was a note to self during debugging.

jwpeterson commented:
I think I have addressed all the comments except for the threading one.  Since that isn't part of our milestone, I'd kind of like to do it on a separate PR.  I am working on it now.

moosebuild commented:
Results of testing f1302bd01457ede3f10f10e36cba1a6c489bafe9 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10983

moosebuild commented:
Results of testing f1302bd01457ede3f10f10e36cba1a6c489bafe9 using moose_PR_test recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10984

jwpeterson commented:
> Failed on: linux-gnu

This was an expect_err failure in parallel.  I don't see this on my system so I'm invalidating and trying again.  This test may need to be restricted to run in serial.

moosebuild commented:
Results of testing f1302bd01457ede3f10f10e36cba1a6c489bafe9 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10984

moosebuild commented:
Results of testing f1302bd01457ede3f10f10e36cba1a6c489bafe9 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/10985

friedmud commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -1786,16 +1794,49 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
+          // sneakily insert bc->variable.number() to make the logic
+          // below a bit simpler.
+          std::set<unsigned int> involved_moose_var_nums;
+          for (unsigned int var=0; var < coupled_moose_vars.size(); ++var)
```
We should be able to figure this out outside of the node loop... it may not take that much time... but on the other had `insert()`ing into a `set` definitely isn't free..

jwpeterson replied:
I tried to figure out how you want to do this, but I couldn't.

Are you wanting to invert the order of the loops, which are currently:
```
for (each boundary node)
  for (each NodalBC)
```

to
 
```
for (each NodalBC)
  for (each boundary node)
```

?

Because whether or not a given variable is coupled to the BC depends on each BC, so the check needs to be inside the BC loop.  The only way the check can be done outside the node loop is if the node loop is inside the BC loop...

friedmud replied:
Before entering the node loop you could do a loop over NodalBCs (all of them) and cache this information for each one.  Then just use that cache inside the nose loop.

This is definitely an optimization... And is lower priority than getting this capability in...

jwpeterson replied:
Ohhh, I see.  Let me see if I can do that...

friedmud commented on framework/src/base/Assembly.C:
```diff
@@ -1389,3 +1389,43 @@ Assembly::addJacobianOffDiagScalar(SparseMatrix<Number> & jacobian, unsigned int
+  // change the capacity of the vectors, so let's be paranoid and
+  // explicitly reserve() the same amount of memory to avoid multiple
+  // push_back() induced allocations.
+  _cached_nodal_bc_rows.reserve(orig_capacity);
```
Once this is threaded... each thread will receive a different amount of nodes each time a threaded loop is called (if you're using TBB).  You may want to give yourself 10% or 20% breathing room here to keep from barely having more rows and getting a re-alloc.

jwpeterson replied:
OK, so you think `_cached_nodal_bc_rows.reserve(1.2*orig_capacity);` is good there?

friedmud replied:
I think that would be fine.  This is at the level of micro-optimization at this point... so without a real use/test case there's no reason to go any further...

friedmud replied:
Wait - you can't just do 1.2*orig_capacity... or it will grow each time.  You need to keep track of the max number of entries you've seen and then set the capacity to 1.2x that.

At this point... screw it.  That's way too complicated for _tiny_ payoff...

jwpeterson replied:
On Tue, Jan 27, 2015 at 11:01 AM, Derek Gaston <notifications@github.com>
wrote:

> > + _cached_nodal_bc_rows.reserve(orig_capacity);
>
> Wait - you can't just do 1.2*orig_capacity... or it will grow each time.
> You need to keep track of the max number of entries you've seen and then
> set the capacity to 1.2x that.
>
In this case orig_capacity is the original size, so it's OK and won't keep
growing.  Bad variable name.

-- 
John

friedmud replied:
Ah - I see.

In that case it doesn't matter anyway then because with threading and TBB it means that the size/capacity is going to move up and down every call on every thread... the only way to avoid that is to keep track of the max this thread has seen and use that the set the capacity each time.

In summary... a 1.2x multiplier for a bit of safety will be just fine ;-)

moosebuild commented:
Results of testing f2f92375ec44f91fdf066971cecbd65e06aff2a4 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/11066

jwpeterson commented:
The variable coupling information for each BC is now cached outside the node loop.

moosebuild commented:
Results of testing f2f92375ec44f91fdf066971cecbd65e06aff2a4 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/11067

moosebuild commented:
Results of testing f2f92375ec44f91fdf066971cecbd65e06aff2a4 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/11068
